### PR TITLE
OCPSTRAT-1439,CLID-28,CLID-170: adds feature to mirror helm charts in oc-mirror v2

### PR DESCRIPTION
**Pull request by aguidirh**
# Description

This PR adds the feature to mirror helm charts in oc-mirror v2. 

Jira issues:
[OCPSTRAT-1439](https://issues.redhat.com/browse/OCPSTRAT-1439)
[CLID-28](https://issues.redhat.com/browse/CLID-28)
[CLID-170](https://issues.redhat.com/browse/CLID-170)

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

With the following ImageSetConfiguration
```
apiVersion: mirror.openshift.io/v2alpha1
kind: ImageSetConfiguration
mirror:
  helm:
    repositories:
      - name: podinfo
        url: https://stefanprodan.github.io/podinfo
        charts:
          - name: podinfo
            version: 5.0.0
      - name: sbo
        url: https://redhat-developer.github.io/service-binding-operator-helm-chart/
    # local:
    #  - name: podinfo-local
    #    path: /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/local-test-clid-28/podinfo-5.0.0.tgz
```

Run all flows (mirrorToDisk, diskToMirror and mirrorToMirror)

`mirrorToDisk`
```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/alex-isc/clid-28-v2.yaml file:///home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-28 --v2
```

`diskToMirror`
```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/alex-isc/clid-28-v2.yaml --from file:///home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-28 docker://localhost:6000 --v2
```

`mirrorToMirror`
```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/alex-isc/clid-28-v2.yaml --workspace file:///home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-28 docker://localhost:6000 --v2
```

## Expected Outcome
All flows should finish successfully and the images should be in the target registry as showed below:

```
curl http://localhost:6000/v2/_catalog | jq
{
  "repositories": [
    "redhat-developer/servicebinding-operator",
    "stefanprodan/podinfo"
  ]
}
```

```
curl http://localhost:6000/v2/stefanprodan/podinfo/tags/list | jq
{
  "name": "stefanprodan/podinfo",
  "tags": [
    "5.0.0"
  ]
}
```

```
curl http://localhost:6000/v2/redhat-developer/servicebinding-operator/tags/list | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   801  100   801    0     0   316k      0 --:--:-- --:--:-- --:--:--  391k
{
  "name": "redhat-developer/servicebinding-operator",
  "tags": [
    "sha256-16286ac84ddd521897d92472dae857a4c18479f255b725dfb683bc72df6e0865",
    "sha256-30bf7f0f21024bb2e1e4db901b1f5e89ab56e0f3197a919d2bbb670f3fe5223a",
    "sha256-67c2a2502f59fac1e7ded9ed19b59bbd4e50f5559a13978a87ecd2283b81e067",
    "sha256-ac47f496fb7ecdcbc371f8c809fad2687ec0c35bbc8c522a7ab63b3e5ffd90ea",
    "sha256-f79f6999a15534dbe56e658caf94fc4b7afb5ceeb7b49f32a60ead06fbd7c3fc",
    "sha256-e4259939a496f292a31b5e57760196d63a8182b999164d93a446da48c4ea24eb",
    "sha256-e01016cacae84dfb6eaf7a1022130e7d95e2a8489c38d4d46e4f734848e93849",
    "sha256-de1881753e82c51b31e958fcf383cb35b0f70f6ec99d402d42243e595d00c6dd",
    "sha256-69a95c6216ead931e01e4144ae8f4fb7ab35d1f68a14c18f6860a085ccb950f5",
    "sha256-cc5aab01ddd3744510c480eb4f58b834936a833d36bec5c9c13fb40bbb06c663"
  ]
}
```

IDMS and ITMS should be generated as showed below:

```
📄 Generating IDMS file...
2024/10/02 16:10:25  [INFO]   : /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-28/working-dir/cluster-resources/idms-oc-mirror.yaml file created
2024/10/02 16:10:25  [INFO]   : 📄 Generating ITMS file...
2024/10/02 16:10:25  [INFO]   : /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-28/working-dir/cluster-resources/itms-oc-mirror.yaml file created
```

`IDMS`
```
---
apiVersion: config.openshift.io/v1
kind: ImageDigestMirrorSet
metadata:
  name: idms-generic-0
spec:
  imageDigestMirrors:
  - mirrors:
    - localhost:6000/redhat-developer
    source: quay.io/redhat-developer
status: {}
```

`ITMS`
```
---
apiVersion: config.openshift.io/v1
kind: ImageTagMirrorSet
metadata:
  name: itms-generic-0
spec:
  imageTagMirrors:
  - mirrors:
    - localhost:6000/stefanprodan
    source: ghcr.io/stefanprodan
status: {}
```

### CODE REVIEW

File: v2/internal/pkg/helm/interface.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	DownloadIndexFile() (string, error)
+}
+
+type chartDownloader interface {
+	DownloadTo(ref, version, dest string) (string, any, error)
+}
+
+type webClient interface {
+	Get(url string) (resp *http.Response, err error)
+}
```
Comment (sherine-k) [inline v2/internal/pkg/helm/interface.go:24]: nice way to enable mocking for tests!

**Review by sherine-k — COMMENTED** [review]
Hi Alex, 
This PR looks great!  
I added a few comments, mostly for understanding better, and making sure we cannot do anything about the TLS configuration.
For future, can we for helm charts do something similar to operators in the concurrent_workers?
when an image fails to mirror, and it's coming from a helm chart: is it possible to have the info about which helm chart is missing an image in the errors that we save to file?
Thanks!

File: v2/internal/pkg/helm/local_stored_collector_test.go [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+
+			imgs, err := helmCollector.HelmImageCollector(ctx)
+
+			if testCase.expectedError == nil {
+				assert.NoError(t, err)
+			}
+
+			if len(testCase.expectedResult) > 0 {
+				assert.NotEmpty(t, imgs)
+				fmt.Println(assert.ElementsMatch(t, testCase.expectedResult, imgs))
```
[the remark this answers is not in the document]
Comment (aguidirh) [reply v2/internal/pkg/helm/local_stored_collector_test.go:434]: It is not  needed. I will remove it, thanks for letting me know.

File: v2/internal/pkg/helm/new.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+func New(log clog.PluggableLoggerInterface,
+	config v2alpha1.ImageSetConfiguration,
+	opts mirror.CopyOptions,
+	indexDownloader indexDownloader,
+	chartDownloader chartDownloader,
+	httpClient webClient,
+) CollectorInterface {
+	lsc = &LocalStorageCollector{Log: log, Config: config, Opts: opts, Helm: NewHelmOptions()}
```
[the remark this answers is not in the document]
Comment (aguidirh) [reply v2/internal/pkg/helm/new.go:18]: New is only called once in the executor.go so there is no chance of it getting overridden currently in runtime.

**Comment by lmzuccarelli** [comment]
@aguidirh  Executed against sonarqube - code coverage is at 74.6% - but its no big deal, as overall we are still at 80% (which meets our agreed AC on code coverage)

Cognitive complexity is high , but honestly in all our logic the value has always been high.  Overall looks good

I'm happy to give it a LGTM

**Pull request merged**